src/studio-home/tabs-section/courses-tab/index.tsx

```diff
@@ -1,67 +1,201 @@
-import { useIntl } from '@edx/frontend-platform/i18n';
+import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';
 import {
   Icon,
   Row,
   Pagination,
   Alert,
   Button,
+  Form,
+  Chip,
 } from '@openedx/paragon';
 import { Error } from '@openedx/paragon/icons';
 
 import { COURSE_CREATOR_STATES } from '@src/constants';
-import { getStudioHomeData, getStudioHomeCoursesParams } from '@src/studio-home/data/selectors';
+import { getStudioHomeData, getStudioHomeCoursesParams, getLoadingStatuses } from '@src/studio-home/data/selectors';
 import { resetStudioHomeCoursesCustomParams, updateStudioHomeCoursesCustomParams } from '@src/studio-home/data/slice';
 import { fetchStudioHomeData } from '@src/studio-home/data/thunks';
 import CardItem from '@src/studio-home/card-item';
 import CollapsibleStateWithAction from '@src/studio-home/collapsible-state-with-action';
 import ProcessingCourses from '@src/studio-home/processing-courses';
 import { LoadingSpinner } from '@src/generic/Loading';
 import AlertMessage from '@src/generic/alert-message';
+import { RequestStatus } from '@src/data/constants';
+import { useMigrationInfo } from '@src/studio-home/data/apiHooks';
 import messages from '../messages';
 import CoursesFilters from './courses-filters';
 import ContactAdministrator from './contact-administrator';
 import './index.scss';
 
-interface Props {
-  coursesDataItems: {
-    courseKey: string;
-    displayName: string;
-    lmsLink: string | null;
-    number: string;
-    org: string;
-    rerunLink: string | null;
-    run: string;
-    url: string;
-  }[];
-  showNewCourseContainer: boolean;
-  onClickNewCourse: () => void;
-  isShowProcessing: boolean;
+interface CardListProps {
+  currentPage: number;
+  handlePageSelected: (page: any) => void;
+  handleCleanFilters: () => void;
+  onClickCard?: (courseId: string) => void;
   isLoading: boolean;
-  isFailed: boolean;
-  numPages: number;
-  coursesCount: number;
+  isFiltered: boolean;
+  hasAbilityToCreateCourse?: boolean;
+  showNewCourseContainer?: boolean;
+  onClickNewCourse?: () => void;
+  inSelectMode?: boolean;
+  selectedCourseId?: string;
+  currentLibraryId?: string;
 }
 
-const CoursesTab: React.FC<Props> = ({
-  coursesDataItems,
-  showNewCourseContainer,
-  onClickNewCourse,
-  isShowProcessing,
+const CardList = ({
+  currentPage,
+  handlePageSelected,
+  handleCleanFilters,
+  onClickCard,
   isLoading,
-  isFailed,
-  numPages = 0,
-  coursesCount = 0,
+  isFiltered,
+  hasAbilityToCreateCourse = false,
+  showNewCourseContainer = false,
+  onClickNewCourse = () => {},
+  inSelectMode = false,
+  selectedCourseId,
+  currentLibraryId,
+}: CardListProps) => {
+  const {
+    courses,
+    numPages,
+    optimizationEnabled,
+  } = useSelector(getStudioHomeData);
+
+  const {
+    data: migrationInfoData,
+  } = useMigrationInfo(courses?.map(item => item.courseKey) || [], currentLibraryId !== undefined);
```

Comment:
I do think it's a bit confusing that this component is located in `studio-home` code, but this code is for the library import page.

If this widget is just a list of courses, it shouldn't need to "know" about the migration status; most "list of courses" use cases don't have anything to do with that. I think it would be much better if we can inject a `<MigrationStatus />` widget, something like

```tsx
<CourseCardList extraWidget={MigrationStatus} ... />
```

and then it can render `<MigrationStatus courseId=X allVisibleCourseIds={[....]} />` - passing the IDs of all the courses in the list so that each MigrationStatus can call useMigrationInfo with the whole list rather than loading each one as a separate REST API call.

What do you think?

Reply:
Thanks for the recommendation!

You are right, I refactored the code in https://github.com/openedx/frontend-app-authoring/pull/2567/commits/f151869da1ad2f4ea9924b490aa1e44ff5ef2f3b